File: ClaudeCodeLearn/lsp_test/calculator.py
"""计算器模块 - 用于测试 LSP 功能"""

from typing import List, Optional


class Calculator:
    """简单计算器类"""

    def __init__(self, name: str = "Calculator"):
        self.name = name
        self.history: List[str] = []

    def add(self, a: float, b: float) -> float:
        """加法运算"""
        result = a + b
        self._record(f"{a} + {b} = {result}")
        return result

    def subtract(self, a: float, b: float) -> float:
        """减法运算"""
        result = a - b
        self._record(f"{a} - {b} = {result}")
        return result

    def multiply(self, a: float, b: float) -> float:
        """乘法运算"""
        result = a * b
        self._record(f"{a} * {b} = {result}")
        return result

    def divide(self, a: float, b: float) -> Optional[float]:
        """除法运算"""
        if b == 0:
            self._record(f"{a} / {b} = Error (division by zero)")
            return None
        result = a / b
        self._record(f"{a} / {b} = {result}")
        return result

    def _record(self, operation: str) -> None:
        """记录操作历史"""
        self.history.append(operation)

    def get_history(self) -> List[str]:
        """获取历史记录"""
        return self.history.copy()

    def clear_history(self) -> None:
        """清空历史记录"""
        self.history.clear()


def create_calculator(name: str) -> Calculator:
    """工厂函数: 创建计算器实例"""
    return Calculator(name)


PI = 3.14159
E = 2.71828

File: ClaudeCodeLearn/lsp_test/calculator.cpp
#include "calculator.hpp"
#include <sstream>

namespace math {

Calculator::Calculator(const std::string& name) : name_(name) {}

double Calculator::add(double a, double b) {
    double result = a + b;
    std::ostringstream oss;
    oss << a << " + " << b << " = " << result;
    record(oss.str());
    return result;
}

double Calculator::subtract(double a, double b) {
    double result = a - b;
    std::ostringstream oss;
    oss << a << " - " << b << " = " << result;
    record(oss.str());
    return result;
}

double Calculator::multiply(double a, double b) {
    double result = a * b;
    std::ostringstream oss;
    oss << a << " * " << b << " = " << result;
    record(oss.str());
    return result;
}

std::optional<double> Calculator::divide(double a, double b) {
    std::ostringstream oss;
    if (b == 0.0) {
        oss << a << " / " << b << " = Error (division by zero)";
        record(oss.str());
        return std::nullopt;
    }
    double result = a / b;
    oss << a << " / " << b << " = " << result;
    record(oss.str());
    return result;
}

void Calculator::record(const std::string& operation) {
    history_.push_back(operation);
}

const std::vector<std::string>& Calculator::getHistory() const {
    return history_;
}

void Calculator::clearHistory() {
    history_.clear();
}

std::string Calculator::getName() const {
    return name_;
}

Calculator createCalculator(const std::string& name) {
    return Calculator(name);
}

} // namespace math
